app/assets/javascripts/libs/input.js

```diff
@@ -444,16 +464,27 @@ export class InputMouse {
   }
 
   pinchStart = (evt: HammerJsEvent) => {
-    this.lastScale = 1;
-
+    this.lastScale = evt.scale;
     // Save position so we can zoom to the pinch start position
-    this.position = this.getRelativeMousePosition({ pageX: evt.center.x, pageY: evt.center.y });
+    // Calculate gesture center ourself as there is a bug in the HammerJS calculation
+    this.position = this.getRelativeMousePosition({
+      pageX: (evt.pointers[0].pageX + evt.pointers[1].pageX) / 2,
+      pageY: (evt.pointers[0].pageY + evt.pointers[1].pageY) / 2,
+    });
   };
 
   pinch = (evt: HammerJsEvent): void => {
-    const delta = evt.scale - this.lastScale;
-    this.lastScale = evt.scale;
-    this.trigger("pinch", 10 * delta);
+    // Abort pinch gesture if another finger is added to the gesture
+    if (evt.pointers.length > 2) this.pinchEnd();
+    if (this.lastScale != null) {
```

Comment:
Is this check only to satisfy flow? Or can it happen that lastScale is undefined at that point? If that's the case, I'd move the line `this.lastScale = evt.scale;` outside of the if-body.

Reply:
Yes it can be undefined as `this.pinchEnd()` sets it to null if the number of fingers used for this gesture exceeds 2 (line above). In that case the pinch gesture should be seen as aborted as the user intends to do something else, so `this.lastScale = evt.scale` should remain in the if-body.
Setting `lastScale` to null effectively ends pinching until the next pinch gesture as `lastScale` is only reset to a non null value in the `pinchStart` method.
Is that understandable?

Reply:
Ok, I understand. Is `pinch` called even after `pinchEnd` was called (for the current touch interaction)? If not, wouldn't the following work:

```
if (evt.pointers.length > 2) {
  this.pinchEnd();
  return;
}
const delta = evt.scale - this.lastScale;
// ...
```

That would be easier to grasp in my opinion, since the logic does not depend on the `lastScale` variable. If the pinch method is still called after pinchEnd was called, it's probably due to hammerjs, and we can't do much about it, right?

Reply:
The problem is that there is no real `pinchend` event in this specific case. We're simply calling our `pinchEnd` method there.
For hammerJS a pinch can be done with two or more fingers. We want it to only be possible with two fingers. So once another finger comes into the game ^^, we artificially abort the pinch on our end and do not react to the following pinch events (that hammerjs sends until hammerjs determines the pinch gesture to be over) :)

Reply:
Alright, I get it :)